server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java

```diff
@@ -868,31 +868,57 @@ public CompletableFuture<Revision> destroyToken(Author author, String appId) {
         requireNonNull(author, "author");
         requireNonNull(appId, "appId");
 
-        // Remove the token from every project.
+        return tokenRepo.push(INTERNAL_PROJECT_DOGMA, Project.REPO_DOGMA, author,
+                              "Delete the token: " + appId,
+                              () -> tokenRepo
+                                      .fetch(INTERNAL_PROJECT_DOGMA, Project.REPO_DOGMA, TOKEN_JSON)
+                                      .thenApply(tokens -> {
+                                          final Token token = tokens.object().get(appId);
+                                          final JsonPointer deletionPath =
+                                                  JsonPointer.compile("/appIds" + encodeSegment(appId) +
+                                                                      "/deletion");
+                                          final Change<?> change = Change.ofJsonPatch(
+                                                  TOKEN_JSON,
+                                                  asJsonArray(new TestAbsenceOperation(deletionPath),
+                                                              new AddOperation(deletionPath,
+                                                                               Jackson.valueToTree(
+                                                                                       UserAndTimestamp.of(
+                                                                                               author)))));
+                                          return HolderWithRevision.of(change, tokens.revision());
+                                      }));
+    }
+
+    /**
+     * Purges the {@link Token} of the specified {@code appId} that was removed before.
+     */
+    public CompletableFuture<Revision> purgeToken(Author author, String appId) {
+        requireNonNull(author, "author");
+        requireNonNull(appId, "appId");
+
         final Collection<Project> projects = new SafeProjectManager(projectManager).list().values();
-        final CompletableFuture<?>[] futures = new CompletableFuture<?>[projects.size()];
-        int i = 0;
-        for (final Project p : projects) {
-            futures[i++] = removeToken(p.name(), author, appId, true).toCompletableFuture();
+
+        // Remove the token from every project.
+        for (Project project : projects) {
+            removeToken(project.name(), author, appId, true).join();
```

Comment:
The original `removeToken` tries to commit to all repositories regardless of the existence of the token in the metadata repo. It may be an inevitable choice because of some race conditions.

Now, token deletion is executed by a scheduler in a leader node. As a result, we can read tokens and then perform deletion only to valid repositories without concurrency issues.
Should we fix `removeToken` as well to reflect on the idea?